lib/swig/GeNNModel.py

```diff
@@ -0,0 +1,934 @@
+            self._slm = slm.SharedLibraryModel_d()
+            self._npType = np.float64
+        elif precision == 'long double':
+            gennFloatType = 'GENN_LONG_DOUBLE'
+            self._slm = slm.SharedLibraryModel_ld()
+            self._npType = np.float128
+        else:
+            raise ValueError( 'Supported precisions are "{0}", '
+                              'but {1} was given'.format( 
+                                  'float, double and long double',
+                                  precision ) )
+        
+        self._built = False
+        self._cpuOnly = cpuOnly
+        if not cpuOnly:
+            self._localhost = pg.init_cuda_mpi()
+        else:
+            self._localhost = 0
+
+        
+        pg.setDefaultVarMode( pg.VarMode_LOC_HOST_DEVICE_INIT_DEVICE )
+        pg.cvar.debugCode = enableDebug
+        pg.cvar.autoInitSparseVars = autoInitSparseVars
+        pg.cvar.buildSharedLibrary = True
+        self._model = pg.NNmodel()
+        self._model.setPrecision( getattr(pg, gennFloatType ) )
+        self.modelName = modelName
+        self.neuronPopulations = {}
+        self.synapsePopulations = {}
+        self.currentSources = {}
+        self._dT = 0.1
+
+    @property
+    def modelName( self ):
+        """Name of the model"""
+        return self._modelName
+
+    @modelName.setter
+    def modelName( self, modelName ):
+        if self._built:
+            raise Exception("GeNN model already built")
+        self._modelName = modelName
+        self._model.setName( modelName )
+
+    @property
+    def t( self ):
+        """Simulation time in ms"""
+        return self._T[0]
+
+    def timestep( self ):
+        """Simulation time step"""
+        return self._TS[0]
+
+    @property
+    def dT( self ):
+        """Step sise"""
+        return self._dT
+
+    @dT.setter 
+    def dT( self, dt ):
+        if self._built:
+            raise Exception("GeNN model already built")
+        self._dT = dt
+        self._model.setDT( dt )
+    
+    def addNeuronPopulation( self, popName, numNeurons, neuron,
+                             paramSpace, varSpace ):
+        """Add a neuron population to the GeNN model
+
+        Args:
+        popName     -- name of the new population
+        numNeurons  -- number of neurons in the new population
+        neuron      -- type of the NeuronModels class as string or instance of neuron class
+                        derived from NeuronModels::Custom class.
+                        sa createCustomNeuronClass
+        paramSpace  -- dict with param values for the NeuronModels class
+        varSpace    -- dict with initial variable values for the NeuronModels class
+        """
+        if self._built:
+            raise Exception("GeNN model already built")
+        if popName in self.neuronPopulations:
+            raise ValueError( 'neuron population "{0}" already exists'.format( popName ) )
+       
+        nGroup = NeuronGroup( popName )
```

Comment:
This applies to all three types of group but I don't quite understand the rationale here. How come you re-implement all the properties of the C++ ``NeuronGroup`` in Python and then immediately duplicate them into a newly created C++ ``NeuronGroup``? Is this so you can defer the creation of the C++ object in future? If not, can't you just call into the wrapped C++ objects?